samples/gdb/mat_pretty_printer.py

```diff
@@ -0,0 +1,255 @@
+    CV_64FC3 = make_type(Depth.CV_64F, 3)
+    CV_64FC4 = make_type(Depth.CV_64F, 4)
+    # CV_64FC(n) = make_type(Depth.CV_64F, (n))
+
+    CV_16FC1 = make_type(Depth.CV_16F, 1)
+    CV_16FC2 = make_type(Depth.CV_16F, 2)
+    CV_16FC3 = make_type(Depth.CV_16F, 3)
+    CV_16FC4 = make_type(Depth.CV_16F, 4)
+    # CV_16FC(n) = make_type(Depth.CV_16F, (n))
+
+
+class Flags:
+    def depth(self):
+        return Depth(self.flags & MagicMasks.DEPTH_MASK.value)
+
+    def dtype(self):
+        depth = self.depth()
+        ret = None
+
+        if depth == Depth.CV_8U:
+            ret = (np.uint8, 'uint8_t')
+        elif depth == Depth.CV_8S:
+            ret = (np.int8, 'int8_t')
+        elif depth == Depth.CV_16U:
+            ret = (np.uint16, 'uint16_t')
+        elif depth == Depth.CV_16S:
+            ret = (np.int16, 'int16_t')
+        elif depth == Depth.CV_32S:
+            ret = (np.int32, 'int32_t')
+        elif depth == Depth.CV_32F:
+            ret = (np.float32, 'float')
+        elif depth == Depth.CV_64F:
+            ret = (np.float64, 'double')
+        elif depth == Depth.CV_16F:
+            ret = (np.float16, 'float16')  # TODO: fix
+
+        return ret
+
+    def type(self):
+        return Type(self.flags & MagicMasks.TYPE_MASK.value)
+
+    def channels(self):
+        return ((self.flags & (511 << 3)) >> 3) + 1
+
+    def is_continuous(self):
+        return (self.flags & MagicValues.CONTINUOUS_FLAG.value) != 0
+
+    def is_submatrix(self):
+        return (self.flags & MagicValues.SUBMATRIX_FLAG.value) != 0
+
+    def __init__(self, flags):
+        self.flags = flags
+
+    def __iter__(self): 
+        return iter({
+                        'type': stri(self.type().name),
+                        'is_continuous': booli(self.is_continuous()),
+                        'is_submatrix': booli(self.is_submatrix())
+                    }.items())
+
+
+class Size(object):
+    def __init__(self, ptr):
+        self.ptr = ptr
+
+    def dims(self):
+        return int((self.ptr - 1).dereference())
+
+    def to_numpy(self):
+        return np.array(list(map(int, to_list(self.ptr, range(self.dims())))), dtype=np.int64)
+
+    def total(self):
+        return np.prod(self.to_numpy())
+
+    def __iter__(self):
+        return iter({'size': stri(self.to_numpy())}.items())
+
+
+def to_list(ptr, it):
+    return [(ptr + i).dereference() for i in it]
+
+
+class Mat:
+    def __init__(self, m, size, flags):
+        (dtype, ctype) = flags.dtype()
+        elsize = np.dtype(dtype).itemsize
+
+        dataptr = int(m['data'])
+        length = (int(m['dataend']) - dataptr) // elsize
+        start = (int(m['datastart']) - dataptr) // elsize
+
+        ptr = m['data']
+
+        if dtype != np.float16:
+            ctype = gdb.lookup_type(ctype)
+            ptr = ptr.cast(ctype.array(length - 1).pointer()).dereference()
+            self.mat = np.array([ptr[i] for i in range(length)], dtype=dtype)
```

Comment:
probably, we iterate over all items in `ptr` here? Are there any scenarios when length != len(ptr) ?

Reply:
Yes, we do iterate over all items in ptr. But ptr is not iterable.

Reply:
Got it, thanks